fix(sem): error with `var`/`lent`-returning templates

zerbina opened the pull request:
## Summary

Fix using templates and macros with a `var T` or `lent T` return type
leading to compiler crashes or wrong code being generated. Both the C
and the VM backend were affected.

## Details

When the receiving type is a `var` or `lent` type, `sigmatch` produced
implicit conversions to said type. These are wrong at the type-level,
as a conversion cannot turn a non-view value into a view. Due a
separate issue with `semAfterMacroCall`, this led to semantically
invalid expressions being produced after expansion of the
aforementioned templates.

The `var` and `lent` modifiers are now dropped from the produced
conversions, both fixing the concrete problem with templates and
allowing for the removal of several workarounds in the compiler. 

Reporting a warning for narrowing object conversions where information
is lost relied on the `var` modifier in conversion. The test is
replaced with looking for the problematic conversion in receiver
contexts (except for `discard`, as the value isn't used there).

Finally, a general test is added for ensuring that templates returning
`var` or `lent` types behave the same as a normal procedure would. They
currently don't, but most of the test at least compiles and runs now.

zerbina committed:
sigmatch: don't produce improper conversions

Going from a non-`var` to a `var` type is not possible via a
conversion.

zerbina committed:
semexprs: cleanup `passToVarParam`

No more special handling for `var` conversions is required (as they no
longer exist).

zerbina committed:
tests: add a general test for `var`/`lent` returning templates

zerbina committed:
remove `var` conversion related workarounds

With the `var` conversion issue resolved, the following workarounds
are now obsolete:
* `cgirgen.wrapInHiddenAddr` doesn't have to special case conversion
  anymore
* `var`/`lent` types needing to be skipped in `vmgen.genObjConv`

zerbina committed:
sempass2: fix the `ImplicitObjConv` false-positives

The previous check relied on the conversion to use a `var` type.
Since this is no longer the case, the check has to performed only on
hidden sub-conversions where the result is read as whole (e.g., passed
to a parameter).

saem commented on compiler/backend/cgirgen.nim:
```diff
@@ -691,22 +691,13 @@ proc wrapInHiddenAddr(cl: TranslateCl, n: CgNode): CgNode =
+  if n.typ.skipTypes(abstractInst).kind != tyVar:
+    newOp(cnkHiddenAddr, n.info, makeVarType(cl.owner, n.typ, cl.idgen), n)
+  else:
+    # XXX: is this case ever reached? It should not be. Raw ``var`` values
+    #      must never be passed directly to ``var`` parameters at the MIR
+    #      level
```
I'm guessing you didn't feel comfortable making this an `unreachable` just yet?

zerbina replied:
Yep, I believe that there are still some cases where this is violated that the test suite doesn't catch.

saem commented on tests/lang_callable/varres/ttemplate_var_lent_res.nim:
```diff
@@ -0,0 +1,47 @@
+  borrowVar(v) = (x: 1)
+  doAssert borrowVar(v) == (x: 1)
+  # test with field write and read:
+  borrowVar(v).x = 2
```
mostly a note for myself, but having this work by clobbering the type is a bit mind bending and possibly dangerous.

zerbina replied:
What do you mean by "clobbering the type"?

saem replied:
If i'm following the expansion correctly, `borrowVar(v)` is just `v`. `v`'s type is `var (,int)`, but by way of the template it's `var tuple`, which due to assignment(?) is treated as `var (, x: int)`, allowing for `borrowVar(v).x`, even though `v` originally didn't have an `x` field.

zerbina replied:
Ah, that you mean. Unnamed tuples are currently implicitly convertible to named ones that have the same structure, but the type of `v` stays the same. The expanded code essentially looks like this:
```nim
Tuple(v).x = 2
```

Relying on this behaviour is not necessary for the test, so I've now changed `v` to be of type `Tuple`.

saem reviewed (approved):
one non-blocking question, otherwise good to go.

zerbina committed:
tests: don't rely on uncertain semantics

A unnamed tuple was substituted in a position a named one is expected,
which currently works, but it's unclear whether it will continue to
do so in the future.

zerbina commented:
Thank you for the review, @saem.

saem commented:
/merge
